File: purview-core/src/main/scala/org/purview/core/transforms/JTransforms.scala
package org.purview.core.transforms

import edu.emory.mathcs.jtransforms._
import org.purview.core.data.Complex
import org.purview.core.data.ImmutableMatrix
import org.purview.core.data.Matrix

case class JTransformsForwardDCT1D() extends Function1[Seq[Float], Seq[Float]] {

  def apply(data: Seq[Float]): Seq[Float] = {
    var array = data.toArray
    val dct = new edu.emory.mathcs.jtransforms.dct.FloatDCT_1D(data.length)
    dct.forward(array, 0, false)
    array.toSeq
  }
}

case class JTransformsInverseDCT1D() extends Function1[Seq[Float], Seq[Float]] {

  def apply(data: Seq[Float]): Seq[Float] = {
    var array = data.toArray
    val dct = new edu.emory.mathcs.jtransforms.dct.FloatDCT_1D(data.length)
    dct.inverse(array, 0, false)
    array.toSeq
  }

}

case class JTransformsForwardDCT2D() extends Function1[Matrix[Float], Matrix[Float]] {

  def apply(data: Matrix[Float]): Matrix[Float] = {
    var array = data.toArray
    val dct = new edu.emory.mathcs.jtransforms.dct.FloatDCT_2D(data.width, data.height)
    dct.forward(array, false)
    new ImmutableMatrix[Float](data.width, data.height, array)
  }

}

case class JTransformsInverseDCT2D() extends Function1[Matrix[Float], Matrix[Float]] {

  def apply(data: Matrix[Float]): Matrix[Float] = {
    var array = data.toArray
    val dct = new edu.emory.mathcs.jtransforms.dct.FloatDCT_2D(data.width, data.height)
    dct.inverse(array, false)
    new ImmutableMatrix[Float](data.width, data.height, array)
  }

}

//case class JTransformsForwardFFT1D() extends Function1[Seq[Float], Seq[Float]] {
//
//  def apply(data: Seq[Float]): Seq[Complex[Float]] = {
//    var array = data.toArray
//    val fft = new edu.emory.mathcs.jtransforms.fft.FloatFFT_1D(data.length)
//    fft.realForwardFull(array, 0)
////    (for(x : Int <- 0 until array.length) yield {
////
////    })
//  }
//}

case class JTransformsInverseFFT1D() extends Function1[Seq[Float], Seq[Float]] {

  def apply(data: Seq[Float]): Seq[Float] = {
    var array = data.toArray
    val dct = new edu.emory.mathcs.jtransforms.dct.FloatDCT_1D(data.length)
    dct.inverse(array, 0, false)
    array.toSeq
  }

}

File: purview-core/src/main/java/org/purview/core/transforms/JTransforms.java
package org.purview.core.transforms;

import edu.emory.mathcs.jtransforms.dct.FloatDCT_1D;
import edu.emory.mathcs.jtransforms.dct.FloatDCT_2D;

class JTransforms {

    public JTransforms() {
    }

    public float[] DCT1DForward(float[] input, boolean scale) {
        float[] data = input.clone();
        FloatDCT_1D dct1d = new FloatDCT_1D(data.length);
        dct1d.forward(data, scale);
        return data;
    }

    public float[] DCT1DInverse(float[] input, boolean scale) {
        float[] data = input.clone();
        FloatDCT_1D dct1d = new FloatDCT_1D(data.length);
        dct1d.inverse(data, scale);
        return data;
    }

    public float[][] DCT2DForward(float[][] input, boolean scale) {
        float[][] data = input.clone();
        FloatDCT_2D dct2d = new FloatDCT_2D(data.length, data[0].length);
        dct2d.forward(data, scale);
        return data;
    }

    public float[][] DCT2DInverse(float[][] input, boolean scale) {
        float[][] data = input.clone();
        FloatDCT_2D dct2d = new FloatDCT_2D(data.length, data[0].length);
        dct2d.inverse(data, scale);
        return data;
    }
}
